internal/unionstore/memdb_test.go

```diff
@@ -1364,3 +1364,230 @@ func TestSnapshotReaderWithWrite(t *testing.T) {
 	check(newRbtDBWithContext(), 48)
 	check(newArtDBWithContext(), 48)
 }
+
+func TestBatchedSnapshotIter(t *testing.T) {
+	check := func(db *artDBWithContext, num int) {
+		// Insert test data
+		for i := 0; i < num; i++ {
+			db.Set([]byte{0, byte(i)}, []byte{0, byte(i)})
+		}
+		h := db.Staging()
+		defer db.Release(h)
+
+		// Create iterator - should be positioned at first key
+		iter := db.BatchedSnapshotIter([]byte{0, 0}, []byte{0, 255}, false)
+		defer iter.Close()
+
+		// Should be able to read first key immediately
+		require.True(t, iter.Valid())
+		require.Equal(t, []byte{0, 0}, iter.Key())
+
+		// Write additional data
+		db.Set([]byte{0, byte(num)}, []byte{0, byte(num)})
+		for i := 0; i < num; i++ {
+			db.Set([]byte{1, byte(i)}, []byte{1, byte(i)})
+		}
+
+		// Verify iteration
+		i := 0
+		for ; i < num; i++ {
+			require.True(t, iter.Valid())
+			require.Equal(t, []byte{0, byte(i)}, iter.Key())
+			require.Equal(t, []byte{0, byte(i)}, iter.Value())
+			require.NoError(t, iter.Next())
+		}
+		require.False(t, iter.Valid())
+	}
+
+	checkReverse := func(db *artDBWithContext, num int) {
+		for i := 0; i < num; i++ {
+			db.Set([]byte{0, byte(i)}, []byte{0, byte(i)})
+		}
+		h := db.Staging()
+		defer db.Release(h)
+
+		iter := db.BatchedSnapshotIter([]byte{0, 0}, []byte{0, 255}, true)
+		defer iter.Close()
+
+		// Should be positioned at last key
+		require.True(t, iter.Valid())
+		require.Equal(t, []byte{0, byte(num - 1)}, iter.Key())
+
+		db.Set([]byte{0, byte(num)}, []byte{0, byte(num)})
+		for i := 0; i < num; i++ {
+			db.Set([]byte{1, byte(i)}, []byte{1, byte(i)})
+		}
+
+		i := num - 1
+		for ; i >= 0; i-- {
+			require.True(t, iter.Valid())
+			require.Equal(t, []byte{0, byte(i)}, iter.Key())
+			require.Equal(t, []byte{0, byte(i)}, iter.Value())
+			require.NoError(t, iter.Next())
+		}
+		require.False(t, iter.Valid())
+	}
+
+	// Run size test cases
+	check(newArtDBWithContext(), 3)
+	check(newArtDBWithContext(), 17)
+	check(newArtDBWithContext(), 64)
+
+	checkReverse(newArtDBWithContext(), 3)
+	checkReverse(newArtDBWithContext(), 17)
+	checkReverse(newArtDBWithContext(), 64)
+}
+
+func TestBatchedSnapshotIterEdgeCase(t *testing.T) {
```

Comment:
Do we have case to cover the memdb is changed bettween two `fillbatch` operations and errors shoul be reported?

Reply:
The "SnapshotChange" case covered the situation